Set Accept-Language header on HTTP requests

maartenba opened the pull request:
(relates to https://github.com/NuGet/Home/issues/2934)

The server currently has no information about the client locale. This PR adds the Accept-Language header to outgoing HTTP requests so that the server can optionally return localized error and warning messages.

Design is based on how UserAgent is added to the request. Simplified version would be to just add the current UI culture instead of having a separate class.

@rrelyea @yishaigalatzer @emgarten @joelverhagen

maartenba committed:
Set Accept-Language header on HTTP requests

joelverhagen commented on src/NuGet.Core/NuGet.Protocol.Core.v2/V2Utilities.cs:
```diff
@@ -44,6 +43,11 @@ public static IPackageRepository GetV2SourceRepository(Configuration.PackageSour
                         {
                             httpReq.UserAgent = _userAgent;
+
+                            if (httpReq.Headers.Get(AcceptLanguage.HeaderName).Length == 0)
```
Can `Get` return null?

maartenba replied:
Added null propagation

joelverhagen commented:
`HttpRequestHeaders.AcceptLanguage`? Also, does this respect `--force-english-output` switch on NuGet.exe and Xplat?

joelverhagen commented:
:watch:

maartenba commented:
Pushed updates.

The `--force-english-output` switch sets `CultureInfo.DefaultThreadCurrentUICulture` to invariant culture. This PR checks `CultureInfo.CurrentUICulture`, which will return an empty string, which in turn will not set the header.

joelverhagen commented on src/NuGet.Core/NuGet.Protocol.Core.v2/V2Utilities.cs:
```diff
@@ -44,6 +43,12 @@ public static IPackageRepository GetV2SourceRepository(Configuration.PackageSour
                         {
                             httpReq.UserAgent = _userAgent;
+
+                            if (httpReq.Headers.Get(AcceptLanguage.HeaderName)?.Length == 0 
```
Couldn't `httpReq.Headers.Get(AcceptLanguage.HeaderName)?.Length == 0` more simply be `string.IsNullOrEmpty(httpReq.Headers.Get(AcceptLanguage.HeaderName))`? What's the intent?

maartenba replied:
Ugh, quick fix from smaller device. Updated. Thanks :)

joelverhagen commented:
LGTM :shipit:

emgarten commented on src/NuGet.Core/NuGet.Protocol.Core.Types/AcceptLanguage.cs:
```diff
@@ -0,0 +1,42 @@
+        static AcceptLanguage()
+        {
+            // Set default Accept-Language header string
+            AcceptLanguageString = CultureInfo.CurrentUICulture.ToString();
```
CurrentCulture is used everywhere else in the client, this should match.

emgarten replied:
Discussed offline with @yishaigalatzer this is the right behavior.

maartenba replied:
👍

emgarten commented:
I think this can be simplified to just updating the httpclient in _CreateHttpClientAsync_. The V2 code path here is only used for the list command and that will go away in the near future. The AcceptLanguage helper class isn't needed.

Caching the locale in a static field will be an issue in Visual Studio where the user can change this through settings. To fully solve this I think the locale would need to be checked for every call since the httpclient is created once per source up front in the VSIX.

maartenba commented:
Updated to remove the cache. Since you suggested to only apply it in `CreateHttpClientAsync`, I removed the `AcceptLanguage` class alltogether as there would be no good use for it. Very thin PR now :-)

emgarten commented on src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs:
```diff
@@ -357,6 +357,10 @@ public HttpSource(PackageSource packageSource, Func<Task<HttpHandlerResource>> m
 
+            // Set accept-language header
+            httpClient.DefaultRequestHeaders.AcceptLanguage.ParseAdd(
+                CultureInfo.CurrentUICulture.ToString());
```
Does this need to check if the string is empty before adding it?

emgarten commented:
Small question on nooping for the empty string, I'm not sure how httpclient handles that. Looks great overall :smile: :shipit:

maartenba commented:
And updated. If it's okay, do I squash it into `dev`? (any other branches?)

emgarten commented:
Squash and put it into dev.

maartenba commented:
Squashed and applied to `dev`. Thanks all for reviewing!
